eip712/src/main/java/pm/gnosis/eip712/EIP712.kt

```diff
@@ -0,0 +1,51 @@
+package pm.gnosis.eip712
+
+import pm.gnosis.crypto.utils.Sha3Utils
+import pm.gnosis.model.Solidity
+import pm.gnosis.model.SolidityBase
+import pm.gnosis.utils.hexToByteArray
+
+val solidityTypes = Solidity.types.keys
+const val DOMAIN_TYPE = "EIP712Domain"
+
+sealed class Type712 {
+    abstract val type: String
+    abstract fun encodeData(): ByteArray
+}
+
+class Literal712(val value: SolidityBase.Type) : Type712() {
+    override fun encodeData(): ByteArray = when (value) {
+        // Solidity.String is also Solidity.Bytes
+        is Solidity.Bytes -> Sha3Utils.keccak(value.items)
+        is SolidityBase.Array<*> -> value.items.map { Literal712(value = it).encodeData() }.reduce { acc, bytes -> acc + bytes }
+        is SolidityBase.Vector<*> -> value.items.map { Literal712(value = it).encodeData() }.reduce { acc, bytes -> acc + bytes }
+        else -> value.encode().hexToByteArray()
+    }
+
+    override val type = value.javaClass.simpleName.toLowerCase()
+}
+
+open class Struct712(override val type: String, val params: List<Pair<String, Type712>>) : Type712() {
+    val typeHash by lazy { Sha3Utils.keccak(encodeType().joinToString(separator = "").toByteArray(charset = Charsets.UTF_8)) }
+
+    override fun encodeData(): ByteArray =
+        params.map { (_, type) ->
+            when (type) {
+                is Struct712 -> type.hashStruct()
+                is Literal712 -> type.encodeData()
+            }
+        }.reduce { acc, bytes -> acc + bytes }
+
+    private fun encodeType(): List<String> {
+        val encodedStruct = params.joinToString(",", prefix = "$type(", postfix = ")", transform = { (name, type) -> "${type.type} $name" })
+        val structParams = params.filter { (_, type) -> !solidityTypes.contains(type.type) }
```

Comment:
This can probably be optimized quite a bit

Reply:
Duplicates can be caught earlier and be skipped (thus removing `.distinct()`. Probably `.sorted()` can also be optimised.

I've written it in this way for readability.